### PR TITLE
[cache] Merge static sliding and static chunked layer

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title. As discussed quite a few times, they are exactly the same, except the Chunked version is more general, as it can handle an arbitrary number of new tokens even after prefill (i.e. prefill caching, chat continuation etc...).
This PR merges them both, to only keep the more general version, which will improve the scope of SlidingWindowLayer usage with the aforementioned use-cases!
Thus Static and Dynamic caches can now be used exactly the same way, in all generality

cc @gante @manueldeprada as well for viz! Finally merging them!

I made sure slow tests on fully sliding (Mistral) and hybrid (Gemma2) are still fine!

### CODE REVIEW

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -504,33 +446,38 @@ def update(
...
-        elif first_cache_position < sliding_window and first_cache_position + query_length > sliding_window:
-            # Here the Cache becomes full with the new input
-            kv_length = first_cache_position + query_length
+        # Not yet full, but becoming full on this update
+        elif self.cumulative_length + query_length > sliding_window:
+            kv_length = self.cumulative_length + query_length
         else:
             # Here the Cache is still smaller than the local size, but we return the local size as it's static
             kv_length = sliding_window
         return kv_length, kv_offset
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py:467]: This method is now much better as we only work with Python scalars, as opposed with a mix of scalar/tensor -> much easier downstream with the masks and compilation

File: src/transformers/cache_utils.py [hunk 2]
```diff
@@ -504,33 +438,38 @@ def update(
...
+        # We only cache the last `sliding_window` tokens
         self.keys.copy_(full_key_states[:, :, -self.max_cache_len :, :])
         self.values.copy_(full_value_states[:, :, -self.max_cache_len :, :])
         # we should return the whole states instead of `self.keys/values` here, as otherwise we lose some context
-        # which is outside the window
         return full_key_states, full_value_states
 
     def get_mask_sizes(self, cache_position: torch.Tensor) -> tuple[int, int]:
         """Return the length and offset of the cache, used to generate the attention mask"""
         query_length = cache_position.shape[0]
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:453]: if you only use the shape we could change this input?

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -1141,7 +1080,7 @@ def __init__(
             if layer_type == "sliding_attention":
                 layer = SlidingWindowLayer(max_cache_len=max_cache_len, sliding_window=config.sliding_window)
             elif layer_type == "chunked_attention":
-                layer = ChunkedSlidingLayer(max_cache_len=max_cache_len, sliding_window=config.attention_chunk_size)
+                layer = SlidingWindowLayer(max_cache_len=max_cache_len, sliding_window=config.attention_chunk_size)
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:1087]: You don't differentiate between chunked (block mask) and sliding because nothing changes for a cache pov! Indeed, they should overall behave the same, mask is what differs!

**Review by ArthurZucker — APPROVED** [review]
i would document why but lgtm

File: src/transformers/cache_utils.py (hunk 3 above)
Comment (Cyrilvallez) [reply src/transformers/cache_utils.py:1087]: Yes, will add a comment to explain here!

File: src/transformers/cache_utils.py (hunk 2 above)
Comment (Cyrilvallez) [reply src/transformers/cache_utils.py:453]: Yes, I'm overall working towards simplifying a lot, and hopefully removing the `cache_position` completely in the future!

File: src/transformers/cache_utils.py [hunk 4]
```diff
@@ -480,16 +401,29 @@ def update(
...
+                # Copy back into `self` (do not just assign again) in order to keep the static dynamo address
+                self.keys.copy_(new_keys)
+                self.values.copy_(new_values)
+                # Very important to return the `self` tensors here, as they have the static dynamo address
                 return self.keys, self.values
-        elif not is_full and cumulative_length + key_states.shape[2] > self.max_cache_len:
+            # Already full but using more than 1 new token (e.g. prefill caching, chat continuation, etc...)
+            else:
+                full_key_states = torch.cat((self.keys[:, :, 1:, :], key_states), dim=-2)
+                full_value_states = torch.cat((self.values[:, :, 1:, :], value_states), dim=-2)
```
Comment (manueldeprada) [inline src/transformers/cache_utils.py:424]: I never understood this `1:` in ChunkedPrefill, maybe worth writing an explanation.

When full and M>1 new tokens arrive, we store last `self.max_cache_len` tokens but we return `self.max_cache_len - 1 + M` tokens (the arrived tokens + `self.keys[:, :, 1:, :]`)

Why this removal of the first token?

Comment (Cyrilvallez) [reply src/transformers/cache_utils.py:424]: It's because the sliding_window is exclusive, i.e. a token can see at most itself and the last `sliding_window - 1` tokens, so the first token of the cache it out-of-scope when we're already full and a new token arrives. Technically, we could cache only `sliding_window - 1` tokens altogether, as I did in `DynamicSlidingWindowLayer`, which would maybe be the preferred way to avoid the additional slicing ops. But it's out-of-scope of this PR, and can technically allow us to rollback one token for some generation methods, so it's not entirely useless

Comment (ArthurZucker) [reply src/transformers/cache_utils.py:424]: we can have a comment to be explicit!

Comment (Cyrilvallez) [reply src/transformers/cache_utils.py:424]: Yes, or we could indeed align with Dynamic version and remove the extra token in the cache, which would be more efficient technically - but this branch is very rare in practice, as it's only used when passing more than 1 token after the cache is already full, which is a specific case of prefix caching or similar
